libcxx/include/__config

```diff
@@ -167,6 +167,11 @@
 // The implementation moved to the header, but we still export the symbols from
 // the dylib for backwards compatibility.
 #    define _LIBCPP_ABI_DO_NOT_EXPORT_TO_CHARS_BASE_10
+// Same memory by providing the allocator more freedom to allocate the most
+// efficient size class by dropping the alignment requirements for std::string's
+// pointer from 16 to 8. This changes the output of std::string::max_size,
+// which makes it ABI breaking
```

Comment:
I think the main source of the ABI break here is that a `std::string`'s `data()` could suddenly be aligned at a 8-byte aligned address and passed to a function that was compiled assuming a 16-bytes alignment? Or am I mistaken?

If it's really *just* about the value of `max_size()`, honestly I think this is something we could probably do unconditionally. I think the likelihood of that actually breaking the ABI for anyone is quite small. Didn't we return an incorrect value from that function for the longest time and it never affected anyone?

Reply:
It is unlikely to "really" break users, in a previous change we saw automated tests failing because 'dynamic libraries'. Theoretically it could cause ODR if you have (inlined / weak) code with different constexpr max_size() compiled in?

But yeah, I think "it should be fine", but I am always a bit.... uncomfortable with these of language lawyer questions :)

Reply:
Ah, I just saw this is actually https://reviews.llvm.org/D155486 IIUC. This provides more background, I see now. Can we abandon that change to clear up our Phabricator queue, or is there stuff left in that patch that we still need to migrate over to Github (or drop).

Coming back to the change itself, FWIW I think I would be happy with going for the "ABI break". Yes this could technically result in ODR violations, but in practice I can hardly imagine someone depending on the exact value of `max_size()`, whose documentation is:

> Returns the maximum number of elements the string is able to hold due to system or library implementation limitations, i.e. `std::distance(begin(), end())` for the largest string.

Reply:
Honestly, I think this looks like a corner case and I think it is unlikely to bite users. Like I said, our value of `max_size()` was broken for the longest time (until 2db67e97712e257d9784f4c3812432443f72145e in 2021).

I would be tempted to make this change unconditionally and give everyone the memory savings. I believe that's the right tradeoff between being paranoid about ABI and actually delivering value to our users.